File: app/src/main/java/com/example/mapsapp/webrtc/webrtc/RTCAudioManager.java
/*
package com.example.mapsapp.webrtc.webrtc;

import android.annotation.SuppressLint;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.media.AudioDeviceInfo;
import android.preference.PreferenceManager;
import android.util.Log;

import org.webrtc.ThreadUtils;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

@SuppressLint("MissingPermission")
public class RTCAudioManager {
    private static final String TAG = RTCAudioManager.class.getSimpleName();
    private static final String SPEAKERPHONE_AUTO = "auto";
    private static final String SPEAKERPHONE_TRUE = "true";
    private static final String SPEAKERPHONE_FALSE = "false";
    private final Context apprtcContext;
    private final String useSpeakerphone;
    private final android.media.AudioManager audioManager;
    private AudioManagerEvents audioManagerEvents;
    private AudioManagerState amState;
    private int savedAudioMode = android.media.AudioManager.MODE_INVALID;
    private boolean savedIsSpeakerPhoneOn = false;
    private boolean savedIsMicrophoneMute = false;
    private boolean hasWiredHeadset = false;
    private AudioDevice defaultAudioDevice;
    private AudioDevice selectedAudioDevice;
    private AudioDevice userSelectedAudioDevice;
    private Set<AudioDevice> audioDevices = new HashSet<>();
    private final BroadcastReceiver wiredHeadsetReceiver;

    private RTCAudioManager(Context context) {
        ThreadUtils.checkIsOnMainThread();
        apprtcContext = context;
        audioManager = ((android.media.AudioManager) context.getSystemService(Context.AUDIO_SERVICE));
        amState = AudioManagerState.UNINITIALIZED;

        useSpeakerphone = PreferenceManager.getDefaultSharedPreferences(context).getString("speakerphone_preference", "auto");
        if (useSpeakerphone.equals(SPEAKERPHONE_FALSE)) {
            defaultAudioDevice = AudioDevice.EARPIECE;
        } else {
            defaultAudioDevice = AudioDevice.SPEAKER_PHONE;
        }

        wiredHeadsetReceiver = new WiredHeadsetReceiver();
    }

    public static RTCAudioManager create(Context context) {
        return new RTCAudioManager(context);
    }

    public void start(AudioManagerEvents audioManagerEvents) {
        ThreadUtils.checkIsOnMainThread();
        if (amState == AudioManagerState.RUNNING) {
            return;
        }

        this.audioManagerEvents = audioManagerEvents;
        amState = AudioManagerState.RUNNING;

        savedAudioMode = audioManager.getMode();
        savedIsSpeakerPhoneOn = audioManager.isSpeakerphoneOn();
        savedIsMicrophoneMute = audioManager.isMicrophoneMute();
        hasWiredHeadset = hasWiredHeadset();

        audioManager.requestAudioFocus(null, android.media.AudioManager.STREAM_VOICE_CALL, android.media.AudioManager.AUDIOFOCUS_GAIN_TRANSIENT);
        audioManager.setMode(android.media.AudioManager.MODE_IN_COMMUNICATION);
        setMicrophoneMute(false);

        userSelectedAudioDevice = AudioDevice.NONE;
        selectedAudioDevice = AudioDevice.NONE;
        audioDevices.clear();

        updateAudioDeviceState();

        registerReceiver(wiredHeadsetReceiver, new IntentFilter(Intent.ACTION_HEADSET_PLUG));
    }

    @SuppressLint("WrongConstant")
    public void stop() {
        ThreadUtils.checkIsOnMainThread();
        if (amState != AudioManagerState.RUNNING) {
            return;
        }
        amState = AudioManagerState.UNINITIALIZED;

        unregisterReceiver(wiredHeadsetReceiver);

        setSpeakerphoneOn(savedIsSpeakerPhoneOn);
        setMicrophoneMute(savedIsMicrophoneMute);
        audioManager.setMode(savedAudioMode);

        audioManager.abandonAudioFocus(null);

        audioManagerEvents = null;
    }

    private void setAudioDeviceInternal(AudioDevice device) {
        switch (device) {
            case SPEAKER_PHONE:
                setSpeakerphoneOn(true);
                break;
            case EARPIECE:
            case WIRED_HEADSET:
                setSpeakerphoneOn(false);
                break;
            default:
                break;
        }
        selectedAudioDevice = device;
    }

    public void setDefaultAudioDevice(AudioDevice defaultDevice) {
        ThreadUtils.checkIsOnMainThread();
        switch (defaultDevice) {
            case SPEAKER_PHONE:
                defaultAudioDevice = defaultDevice;
                break;
            case EARPIECE:
                if (hasEarpiece()) {
                    defaultAudioDevice = defaultDevice;
                } else {
                    defaultAudioDevice = AudioDevice.SPEAKER_PHONE;
                }
                break;
            default:
                break;
        }
        updateAudioDeviceState();
    }

    public void selectAudioDevice(AudioDevice device) {
        ThreadUtils.checkIsOnMainThread();
        if (!audioDevices.contains(device)) {
            return;
        }
        userSelectedAudioDevice = device;
        updateAudioDeviceState();
    }

    public Set<AudioDevice> getAudioDevices() {
        ThreadUtils.checkIsOnMainThread();
        return Collections.unmodifiableSet(new HashSet<>(audioDevices));
    }

    public AudioDevice getSelectedAudioDevice() {
        ThreadUtils.checkIsOnMainThread();
        return selectedAudioDevice;
    }

    private void registerReceiver(BroadcastReceiver receiver, IntentFilter filter) {
        apprtcContext.registerReceiver(receiver, filter);
    }

    private void unregisterReceiver(BroadcastReceiver receiver) {
        apprtcContext.unregisterReceiver(receiver);
    }

    private void setSpeakerphoneOn(boolean on) {
        boolean wasOn = audioManager.isSpeakerphoneOn();
        if (wasOn == on) {
            return;
        }
        audioManager.setSpeakerphoneOn(on);
    }

    private void setMicrophoneMute(boolean on) {
        boolean wasMuted = audioManager.isMicrophoneMute();
        if (wasMuted == on) {
            return;
        }
        audioManager.setMicrophoneMute(on);
    }

    private boolean hasEarpiece() {
        return apprtcContext.getPackageManager().hasSystemFeature(PackageManager.FEATURE_TELEPHONY);
    }

    @Deprecated
    private boolean hasWiredHeadset() {
        @SuppressLint("WrongConstant") final AudioDeviceInfo[] devices = audioManager.getDevices(android.media.AudioManager.GET_DEVICES_ALL);
        for (AudioDeviceInfo device : devices) {
            final int type = device.getType();
            if (type == AudioDeviceInfo.TYPE_WIRED_HEADSET || type == AudioDeviceInfo.TYPE_USB_DEVICE) {
                return true;
            }
        }
        return false;
    }

    public void updateAudioDeviceState() {
        ThreadUtils.checkIsOnMainThread();

        Set<AudioDevice> newAudioDevices = new HashSet<>();

        if (hasWiredHeadset) {
            newAudioDevices.add(AudioDevice.WIRED_HEADSET);
        } else {
            newAudioDevices.add(AudioDevice.SPEAKER_PHONE);
            if (hasEarpiece()) {
                newAudioDevices.add(AudioDevice.EARPIECE);
            }
        }

        boolean audioDeviceSetUpdated = !audioDevices.equals(newAudioDevices);
        audioDevices = newAudioDevices;

        if (hasWiredHeadset && userSelectedAudioDevice == AudioDevice.SPEAKER_PHONE) {
            userSelectedAudioDevice = AudioDevice.WIRED_HEADSET;
        }
        if (!hasWiredHeadset && userSelectedAudioDevice == AudioDevice.WIRED_HEADSET) {
            userSelectedAudioDevice = AudioDevice.SPEAKER_PHONE;
        }

        AudioDevice newAudioDevice = hasWiredHeadset ? AudioDevice.WIRED_HEADSET : defaultAudioDevice;

        if (newAudioDevice != selectedAudioDevice || audioDeviceSetUpdated) {
            setAudioDeviceInternal(newAudioDevice);

            if (audioManagerEvents != null) {
                audioManagerEvents.onAudioDeviceChanged(selectedAudioDevice, audioDevices);
            }
        }
    }

    public enum AudioDevice {
        SPEAKER_PHONE, WIRED_HEADSET, EARPIECE, NONE
    }

    public enum AudioManagerState {
        UNINITIALIZED, PREINITIALIZED, RUNNING,
    }

    public interface AudioManagerEvents {
        void onAudioDeviceChanged(AudioDevice selectedAudioDevice, Set<AudioDevice> availableAudioDevices);
    }

    private class WiredHeadsetReceiver extends BroadcastReceiver {
        private static final int STATE_UNPLUGGED = 0;
        private static final int STATE_PLUGGED = 1;

        @Override
        public void onReceive(Context context, Intent intent) {
            int state = intent.getIntExtra("state", STATE_UNPLUGGED);
            hasWiredHeadset = (state == STATE_PLUGGED);
            updateAudioDeviceState();
        }
    }
}
*/

File: app/src/main/java/com/example/mapsapp/webrtc/webrtc/RTCAudioManager.kt
package com.example.mapsapp.webrtc.webrtc

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.media.AudioDeviceInfo
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.os.Build
import android.util.Log
import org.webrtc.ThreadUtils
import java.util.Collections
import java.util.HashSet

@SuppressLint("MissingPermission")
class RTCAudioManager private constructor(private val context: Context) {

    private val TAG = "RTCAudioManager"
    private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private var audioManagerEvents: AudioManagerEvents? = null

    private var savedAudioMode = AudioManager.MODE_NORMAL
    private var savedIsSpeakerPhoneOn = false
    private var savedIsMicrophoneMute = false
    private var hasWiredHeadset = false

    private var defaultAudioDevice = AudioDevice.SPEAKER_PHONE
    private var selectedAudioDevice = AudioDevice.NONE
    private var userSelectedAudioDevice = AudioDevice.NONE
    private var audioDevices = HashSet<AudioDevice>()

    private val wiredHeadsetReceiver = WiredHeadsetReceiver()

    // 🔴 Eksik olan: Audio Focus Management (Şimdi eklendi)
    private var audioFocusRequest: AudioFocusRequest? = null

    companion object {
        fun create(context: Context): RTCAudioManager {
            return RTCAudioManager(context)
        }
    }

    fun start(audioManagerEvents: AudioManagerEvents) {
        ThreadUtils.checkIsOnMainThread()
        this.audioManagerEvents = audioManagerEvents

        savedAudioMode = audioManager.mode
        savedIsSpeakerPhoneOn = audioManager.isSpeakerphoneOn
        savedIsMicrophoneMute = audioManager.isMicrophoneMute

        hasWiredHeadset = isWiredHeadsetConnected()

        audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
        audioManager.isMicrophoneMute = false

        // 🔴 Eksik olan: Audio Focus Yönetimi (Şimdi eklendi)
        requestAudioFocus()

        userSelectedAudioDevice = AudioDevice.NONE
        selectedAudioDevice = AudioDevice.NONE
        audioDevices.clear()

        updateAudioDeviceState()
        registerReceiver(wiredHeadsetReceiver, IntentFilter(Intent.ACTION_HEADSET_PLUG))
    }

    fun stop() {
        ThreadUtils.checkIsOnMainThread()
        unregisterReceiver(wiredHeadsetReceiver)
        audioManager.isSpeakerphoneOn = savedIsSpeakerPhoneOn
        audioManager.isMicrophoneMute = savedIsMicrophoneMute
        audioManager.mode = savedAudioMode
        abandonAudioFocus()
        audioManagerEvents = null
    }

    private fun setAudioDeviceInternal(device: AudioDevice) {
        selectedAudioDevice = device
        when (device) {
            AudioDevice.SPEAKER_PHONE -> audioManager.isSpeakerphoneOn = true
            AudioDevice.EARPIECE, AudioDevice.WIRED_HEADSET -> audioManager.isSpeakerphoneOn = false
            else -> {}
        }
    }

    fun setDefaultAudioDevice(defaultDevice: AudioDevice) {
        ThreadUtils.checkIsOnMainThread()
        defaultAudioDevice = if (defaultDevice == AudioDevice.EARPIECE && hasEarpiece()) {
            defaultDevice
        } else {
            AudioDevice.SPEAKER_PHONE
        }
        updateAudioDeviceState()
    }

    fun selectAudioDevice(device: AudioDevice) {
        ThreadUtils.checkIsOnMainThread()
        if (audioDevices.contains(device)) {
            userSelectedAudioDevice = device
            updateAudioDeviceState()
        }
    }

    fun getAudioDevices(): Set<AudioDevice> {
        return Collections.unmodifiableSet(HashSet(audioDevices))
    }

    fun getSelectedAudioDevice(): AudioDevice {
        return selectedAudioDevice
    }

    private fun registerReceiver(receiver: BroadcastReceiver, filter: IntentFilter) {
        context.registerReceiver(receiver, filter)
    }

    private fun unregisterReceiver(receiver: BroadcastReceiver) {
        try {
            context.unregisterReceiver(receiver)
        } catch (e: IllegalArgumentException) {
            Log.w(TAG, "Receiver was not registered: ${e.message}")
        }
    }

    private fun updateAudioDeviceState() {
        ThreadUtils.checkIsOnMainThread()
        val newAudioDevices = HashSet<AudioDevice>()

        if (hasWiredHeadset) {
            newAudioDevices.add(AudioDevice.WIRED_HEADSET)
        } else {
            newAudioDevices.add(AudioDevice.SPEAKER_PHONE)
            if (hasEarpiece()) {
                newAudioDevices.add(AudioDevice.EARPIECE)
            }
        }

        if (audioDevices != newAudioDevices) {
            audioDevices = newAudioDevices
            val newAudioDevice = if (hasWiredHeadset) AudioDevice.WIRED_HEADSET else defaultAudioDevice
            setAudioDeviceInternal(newAudioDevice)
            audioManagerEvents?.onAudioDeviceChanged(selectedAudioDevice, audioDevices)
        }
    }

    private fun hasEarpiece(): Boolean {
        return context.packageManager.hasSystemFeature(PackageManager.FEATURE_TELEPHONY)
    }

    private fun isWiredHeadsetConnected(): Boolean {
        val devices = audioManager.getDevices(AudioManager.GET_DEVICES_INPUTS)
        return devices.any { it.type == AudioDeviceInfo.TYPE_WIRED_HEADSET || it.type == AudioDeviceInfo.TYPE_USB_DEVICE }
    }

    private fun requestAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val focusRequest = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
                .setOnAudioFocusChangeListener { }
                .build()
            val result = audioManager.requestAudioFocus(focusRequest)
            if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                Log.d(TAG, "Audio focus granted")
            } else {
                Log.e(TAG, "Audio focus request failed")
            }
            audioFocusRequest = focusRequest
        } else {
            @Suppress("DEPRECATION")
            val result = audioManager.requestAudioFocus(null, AudioManager.STREAM_VOICE_CALL, AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
            if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
                Log.d(TAG, "Audio focus granted (Legacy)")
            } else {
                Log.e(TAG, "Audio focus request failed (Legacy)")
            }
        }
    }

    private fun abandonAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            audioFocusRequest?.let {
                audioManager.abandonAudioFocusRequest(it)
                Log.d(TAG, "Audio focus abandoned")
            }
        } else {
            @Suppress("DEPRECATION")
            audioManager.abandonAudioFocus(null)
            Log.d(TAG, "Audio focus abandoned (Legacy)")
        }
    }

    interface AudioManagerEvents {
        fun onAudioDeviceChanged(selectedAudioDevice: AudioDevice, availableAudioDevices: Set<AudioDevice>)
    }

    enum class AudioDevice {
        SPEAKER_PHONE, WIRED_HEADSET, EARPIECE, NONE
    }

    private inner class WiredHeadsetReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            hasWiredHeadset = intent.getIntExtra("state", 0) == 1
            updateAudioDeviceState()
        }
    }
}
